classes/quiz_settings.php

```diff
@@ -147,6 +152,193 @@ protected static function define_properties() {
                 'default' => '',
                 'null' => NULL_ALLOWED,
             ],
+            'config' => [
+                'type' => PARAM_RAW,
+                'default' => '',
+                'null' => NULL_ALLOWED,
+            ],
+        ];
+    }
+
+    /**
+     * Hook to execute before an update.
+     *
+     * Please note that at this stage the data has already been validated and therefore
+     * any new data being set will not be validated before it is sent to the database.
+     */
+    protected function before_update() {
+        $this->before_save();
+    }
+
+    /**
+     * Hook to execute before a create.
+     *
+     * Please note that at this stage the data has already been validated and therefore
+     * any new data being set will not be validated before it is sent to the database.
+     */
+    protected function before_create() {
+        $this->before_save();
+    }
+
+    /**
+     * As there is no hook for before both create and update, this function is called by both hooks.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     * @throws \coding_exception
+     */
+    private function before_save() {
+        // Recalculate config and config key.
+        $this->compute_config();
+        $this->compute_config_key();
+    }
+
+    /**
+     * Generate the config key from the config string.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     * @throws \coding_exception
+     */
+    private function compute_config_key() {
+        $config = $this->get('config');
+        if (!empty($config)) {
```

Comment:
in what case this could be empty? Should we cover this with tests?

Reply:
Removed conditional and https://github.com/catalyst/moodle-quizaccess_seb/pull/37 adds test for empty config.